Fix duplicate node-drainer events by sorting pod lists

Gyan172004 opened the pull request:
## Summary

**Issue:**  
Issue #339 – The node-drainer emitted multiple Kubernetes events for the same set of pods because the pod list in the event message wasn’t ordered. As a result, Kubernetes treated each permutation of the pod list as a separate event instead of aggregating them.

**Reason:**  
Go’s iteration over maps/slices is non-deterministic. Without sorting the pod names, identical pod sets produced different event message strings, which caused Kubernetes to record duplicate events rather than incrementing counts.

**Solution:**  
Sort the pod names before constructing event messages in both `AllowCompletion` (reconciler.go) and `DeleteAfterTimeout` (informers.go). This ensures the same pod set always generates a consistent event string, allowing Kubernetes to aggregate events correctly.

## Type of Change
- [X] 🐛 Bug fix
- [ ] ✨ New feature
- [ ] 💥 Breaking change
- [ ] 📚 Documentation
- [ ] 🔧 Refactoring
- [ ] 🔨 Build/CI

## Component(s) Affected
- [ ] Core Services
- [ ] Documentation/CI
- [x] Fault Management
- [ ] Health Monitors
- [ ] Janitor
- [ ] Other: ____________

## Testing
- [X] Tests pass locally
- [X] Manual testing completed
- [X] No breaking changes (or documented)

## Checklist
- [X] Self-review completed
- [X] Documentation updated (if needed)
- [X] Ready for review


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Pod lists in status messages and node events are now deterministically and alphabetically sorted for consistent, clearer logging and to prevent spurious or duplicate events.

* **Tests**
  * Integration tests updated to run multiple reconciliation steps, validate sorted pod lists, and cover multi-pod scenarios. A per-test reconciliation count was added so tests can simulate repeated processing and verify stable event messages.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Gyan172004 committed:
fix: ensure deterministic pod event messages

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

Gyan172004 committed:
Add integration test for pod list sorting consistency

Test verifies that sorting prevents duplicate Kubernetes events across reconciliation loops.

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

Gyan172004 committed:
making test more robust,and early break

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

Gyan172004 commented:
Here in the test, I are following an approach to catch non-deterministic behavior:  
The core problem for issue is that the order we get when iterating a map is not fixed, so each time we loop over it, we might* get a different sequence of items.

Our assumption is that out of 5 attempts, the internal map order will change at least once. The initial assumption was 10, but we saw that iterations were failing within 3–4 tries, so we cut it down to 5.

The test runs in a loop, generating the pod event message each time. It takes the first message as the reference and checks that all subsequent messages match. If any sequence is different, it means the bug is present, and the test fails immediately.

To make sure this was foolproof, I ran this test case in a loop 10,000 times with our sorting code removed, and it resulted in all FAIL. This means there were no false positives and shows that the probability of this test missing the bug is <0.0001.

I chose this iterative comparison because trying to write 2–3 separate tests(which llm was also doing same) is not reliable; the map might give the same order by chance, leading to a false positive. 

**Let me know if this makes sense 
Open to Feedback**

@lalitadithya

Gyan172004 committed:
refactor: remove verbose comments from integration test

Signed-off-by: Gyan172004 <gkoncs@gmail.com>

lalitadithya commented on node-drainer/pkg/reconciler/reconciler_integration_test.go:
```diff
@@ -1020,3 +1020,98 @@ func getHistogramCount(t *testing.T, histogram prometheus.Histogram) uint64 {
 }
+
+// TestReconciler_PodListSortingConsistentEvents verifies that pod list sorting prevents duplicate Kubernetes events by ensuring consistent event messages across multiple reconciliation calls.
+func TestReconciler_PodListSortingConsistentEvents(t *testing.T) {
```
can we instead update the existing test to cover this scenario -- https://github.com/NVIDIA/NVSentinel/blob/1ac2e83ae8947f7c6cd66cae2c08b699294eeee0/node-drainer/pkg/reconciler/reconciler_integration_test.go#L166

Gyan172004 replied:
just to confirm, do you mean I should add the `AllowCompletion mode waits for pods to complete` logic into my test, or the other way around?

Gyan172004 replied:
> we can run the test multiple times to confirm.

Yes, running it multiple times will work correctly. 

I initially tried to avoid running it multiple times inside the table-driven test because that would require passing the reconciler as a parameter to the validateFunc, which would then need to be added to all test cases in the table.

Gyan172004 committed:
test: integrate pod sorting verification into existing table-driven test

- Update 'AllowCompletion mode' test to verify consistent pod ordering
- Add numReconciles field to support multiple reconciliation cycles
- Test with 3 pods in non-alphabetical order to validate sorting
- Verify only 1 event created despite 5 reconciliations (deduplication works)
- Remove standalone TestReconciler_PodListSortingConsistentEvents test

Signed-off-by: Gyan172004 <gkoncs@gmail.com>
